subsys/disk/disk_access_usdhc.c

```diff
@@ -0,0 +1,2879 @@
+	struct usdhc_data *data = &priv->op_context.data;
+
+	if (!priv->host_ready) {
+		return -ENODEV;
+	}
+
+	/* reset variables */
+	priv->card_info.card_flags = 0U;
+	/* set DATA bus width 1bit at beginning*/
+	usdhc_set_bus_width(base, USDHC_DATA_BUS_WIDTH_1BIT);
+	/*set card freq to 400KHZ at begging*/
+	priv->card_info.busclk_hz =
+		usdhc_set_sd_clk(base, priv->host_config.src_clk_hz,
+			SDMMC_CLOCK_400KHZ);
+	/* send card active */
+	ret = usdhc_set_sd_active(base);
+	if (ret == false) {
+		return -EIO;
+	}
+
+	/* Get host capability. */
+	usdhc_get_host_capability(base, &priv->host_capability);
+
+	/* card go idle */
+	usdhc_op_ctx_init(priv, 1, SDHC_GO_IDLE_STATE, 0, SDHC_RSP_TYPE_NONE);
+
+	ret = usdhc_xfer(priv);
+	if (ret) {
+		return ret;
+	}
+
+	if (USDHC_SUPPORT_V330_FLAG != SDMMCHOST_NOT_SUPPORT) {
+		app_cmd_41_arg |= (SD_OCR_VDD32_33FLAG | SD_OCR_VDD33_34FLAG);
+		priv->card_info.voltage = SD_VOL_3_3_V;
+	} else if (USDHC_SUPPORT_V300_FLAG != SDMMCHOST_NOT_SUPPORT) {
+		app_cmd_41_arg |= SD_OCR_VDD29_30FLAG;
+		priv->card_info.voltage = SD_VOL_3_3_V;
+	}
+
+	/* allow user select the work voltage, if not select,
+	 * sdmmc will handle it automatically
+	 */
+	if (USDHC_SUPPORT_V180_FLAG != SDMMCHOST_NOT_SUPPORT) {
+		app_cmd_41_arg |= SD_OCR_SWITCH_18_REQ_FLAG;
+	}
+
+	/* Check card's supported interface condition. */
+	usdhc_op_ctx_init(priv, 1, SDHC_SEND_IF_COND,
+		SDHC_VHS_3V3 | SDHC_CHECK, SDHC_RSP_TYPE_R7);
+
+	retry = 10;
+	while (retry) {
+		ret = usdhc_xfer(priv);
+		if (!ret) {
+			if ((cmd->response[0U] & 0xFFU) != SDHC_CHECK) {
+				ret = -ENOTSUP;
+			} else {
+				break;
+			}
+		}
+		retry--;
+	}
+
+	if (!ret) {
+		/* SDHC or SDXC card */
+		app_cmd_41_arg |= SD_OCR_HOST_CAP_FLAG;
+		priv->card_info.card_flags |= USDHC_SDHC_FLAG;
+	} else {
+		/* SDSC card */
+		LOG_ERR("USDHC SDSC not implemented yet!\r\n");
+		return -ENOTSUP;
+	}
+
+	/* Set card interface condition according to SDHC capability and
+	 * card's supported interface condition.
+	 */
+APP_SEND_OP_COND_AGAIN:
+	usdhc_op_ctx_init(priv, 1, 0, 0, SDHC_RSP_TYPE_NONE);
+	ret = usdhc_app_host_cmd(priv, NXP_SDMMC_MAX_VOLTAGE_RETRIES, 0,
+		SDHC_APP_SEND_OP_COND, app_cmd_41_arg,
+		SDHC_RSP_TYPE_R1, SDHC_RSP_TYPE_R3, 1);
+	if (ret) {
+		LOG_ERR("APP Condition CMD failed:%d\r\n", ret);
+		return ret;
+	}
+	if (cmd->response[0U] & SD_OCR_PWR_BUSY_FLAG) {
+		/* high capacity check */
+		if (cmd->response[0U] & SD_OCR_CARD_CAP_FLAG) {
+			priv->card_info.card_flags |= SDHC_HIGH_CAPACITY_FLAG;
+		}
+		/* 1.8V support */
+		if (cmd->response[0U] & SD_OCR_SWITCH_18_ACCEPT_FLAG) {
+			priv->card_info.card_flags |= SDHC_1800MV_FLAG;
+		}
+		priv->card_info.raw_ocr = cmd->response[0U];
+	} else {
+		goto APP_SEND_OP_COND_AGAIN;
+	}
+
+	/* check if card support 1.8V */
+	if ((priv->card_info.card_flags & USDHC_VOL_1_8V_FLAG)) {
+		usdhc_op_ctx_init(priv, 1, SDHC_VOL_SWITCH,
+			0, SDHC_RSP_TYPE_R1);
+
+		ret = usdhc_xfer(priv);
+		if (!ret) {
+			ret = usdhc_
```

Comment:
You can tie DT_NXP_IMX_USDHC_SLOT_0_BASE_ADDRESS to a generic DT_USDHC_SLOT_0_BASE_ADDRESS using dts_fixup.h 
This will help drop nxp ifdefs.

Reply:
Hi, I can't fully understand how to drop nxp ifdefs..

Reply:
@JunYangNXP See zephyr/soc/arm/nxp_imx/rt/dts_fixup.h
Look at the last definition:
#define DT_USBD_MCUX_EHCI_MAXIMUM_SPEED   DT_NXP_KINETIS_USBD_402E0000_MAXIMUM_SPEED

DT_USBD_MCUX_EHCI_MAXIMUM_SPEED - used in code
DT_NXP_KINETIS_USBD_402E0000_MAXIMUM_SPEED - generated from dts. Look at <sample>/build/zephyr/include/generated/generated_dts_board.conf after compiling at the generated macros from dts.

In your case DT_NXP_IMX_USDHC_SLOT_0_BASE_ADDRESS is a generated macro.
Using a generic definition will help reuse the same code without dts hard-coding.

Reply:
For MIMXRT1050EVK board, the generated macro is:
#define DT_NXP_IMX_USDHC_0_BASE_ADDRESS				DT_NXP_IMX_USDHC_402C0000_BASE_ADDRESS
 but the DT_NXP_IMX_USDHC_0_BASE_ADDRESS depends on board design, if customer uses USDHC2 for their board SDHC slot, the generated macro should be:
#define DT_NXP_IMX_USDHC_0_BASE_ADDRESS				DT_NXP_IMX_USDHC_402C4000_BASE_ADDRESS
seems hard to fix this marco.

please correct me if any mistake.

Reply:
The customer will have something like:
#define DT_NXP_IMX_USDHC_>>1<<_BASE_ADDRESS DT_NXP_IMX_USDHC_402C4000_BASE_ADDRESS
This can be added currently in the code.
A custom board would use the same soc configuration, but he would have a new board directory with different pinmux and other cofigurations.

Reply:
DT_NXP_IMX_USDHC_0_BASE_ADDRESS means the SDHC slot 0 (1st) on board, the slot 0 could connect to USDHC0 or USDHC1 which depends on board design.
If USDHC0 is connected, DT_NXP_IMX_USDHC_0_BASE_ADDRESS  is DT_NXP_IMX_USDHC_402C0000_BASE_ADDRESS, otherwise, it's DT_NXP_IMX_USDHC_402C4000_BASE_ADDRESS.

Reply:
DT_NXP_IMX_USDHC_x_BASE_ADDRESS refers to SoC internal device's address, not the external slot on the board.
We use dts to associate slot to soc's device.

Reply:
on mimxrt1050_evk.dts:
I defined:
&usdhc1 {
	status = "ok";
	pwr-gpios = <&gpio1 5 0>;
	cd-gpios = <&gpio2 28 0>;
};
And generate macros:
#define DT_NXP_IMX_USDHC_0_BASE_ADDRESS				DT_NXP_IMX_USDHC_402C0000_BASE_ADDRESS

But if I change the board dst:
&usdhc2 {
	status = "ok";
	pwr-gpios = <&gpio1 5 0>;
	cd-gpios = <&gpio2 28 0>;
};
it generates:
#define DT_NXP_IMX_USDHC_0_BASE_ADDRESS				DT_NXP_IMX_USDHC_402C4000_BASE_ADDRESS

Reply:
You are right, unless both devices have status = "ok", the only device enabled will have 0 set.

You can have multiple dts_fixup.h some in soc, some in boards. At a board level you can have USDHC_0 defined as using &usdhc1 and another board to use &usdhc2.

Just a few boards have dts_fixup.h defined in board folder, as most development boards for a soc family reuse the same board design.

Reply:
Are there other i.MX RT 10xx using a diferent usdhc device for the slot?

Reply:
I did't check other boards, but I think board designer prefer to use USDHC0 unless pins of USDHC0 are muxed for some other purpose.:)